### PR TITLE
Improve test and linting setup

### PR DESCRIPTION
**Pull request by JeroenDeDauw**
I have use for my old interface and test doubles in some WMDE code. So this package can get onto WP ;)

The "Improve test and linting setup" commit contains:

* Updated PHPUnit version
* PHPUnit now installed as dev dependency
* Style checks added
* Makefile added and documented in README
* Travis config updated and simplified
* Code updated to match coding style

A lot of what is in that commit is copied from https://github.com/JeroenDeDauw/new-php-library

### CODE REVIEW

**Comment by JeroenDeDauw** [comment]
It'd be cool if I could get admin on this repo and the packagist entry

**Comment by mwjames** [comment]
> It'd be cool if I could get admin on this repo and the packagist entry

Done.

File: src/MessageReporterAwareTrait.php [hunk 1]
```diff
@@ -16,9 +16,10 @@ trait MessageReporterAwareTrait {
 	protected $messageReporter;
 
 	/**
+	 * @param MessageReporter $messageReporter
+	 *
 	 * @since 1.3
 	 *
-	 * @param MessageReporter $messageReporter
 	 */
```
Comment (mwjames) [inline src/MessageReporterAwareTrait.php]: Not sure why, but `since` comes always before `param`/`return`/`throws` so some linting rules seems off here.

File: tests/bootstrap.php [hunk 2]
```diff
@@ -1,24 +1,16 @@
...
- *
- * @author mwjames
- */
+declare( strict_types = 1 );
 
-error_reporting( E_ALL | E_STRICT );
-date_default_timezone_set( 'UTC' );
-
-if ( PHP_SAPI !== 'cli' && PHP_SAPI !== 'phpdbg' ) {
+if ( PHP_SAPI !== 'cli' ) {
```
Comment (mwjames) [inline tests/bootstrap.php]: `&& PHP_SAPI !== 'phpdbg'` should remain so it can run with the internal PHP debugger

File: tests/phpunit/MessageReporterAwareTest.php [hunk 3]
```diff
@@ -1,41 +0,0 @@
-<?php
```
Comment (mwjames) [inline tests/phpunit/MessageReporterAwareTest.php]: Why is this removed?

File: tests/ci/upload-coverage-report.sh [hunk 4]
```diff
@@ -1,10 +0,0 @@
-#! /bin/bash
```
Comment (mwjames) [inline tests/ci/upload-coverage-report.sh:1]: Removed why?

File: tests/ci/run-tests.sh [hunk 5]
```diff
@@ -1,20 +0,0 @@
-#! /bin/bash
```
Comment (mwjames) [inline tests/ci/run-tests.sh:1]: Removed why?

File: src/MessageReporterAwareTrait.php (hunk 1 above)
Comment (JeroenDeDauw) [reply src/MessageReporterAwareTrait.php]: The CS does not care. For some reason my IDE did that reorder. Fixed now.

Comment (mwjames) [reply src/MessageReporterAwareTrait.php]: I know you are not a big fan of these comments but I find them helpful when something was introduced without having to consult `git blame` etc.

Comment (JeroenDeDauw) [reply src/MessageReporterAwareTrait.php]: Pff it also changed in other files. I'd rather not care about that as hopefully soon this can use PHP 7 and those type hints in the docs can be ditched

Comment (JeroenDeDauw) [reply src/MessageReporterAwareTrait.php]: Fun fact: it was me that introduced those @since comments into MediaWiki world. In the last few years I am no longer adding them within classes with some rare exceptions. The tags on the class level suffice to not need to do git blame.

File: tests/phpunit/MessageReporterAwareTest.php (hunk 3 above)
Comment (JeroenDeDauw) [reply tests/phpunit/MessageReporterAwareTest.php]: Because the test is not testing any production code logic and violates PSR-1 by having two classes in a file

File: tests/ci/upload-coverage-report.sh (hunk 4 above)
Comment (JeroenDeDauw) [reply tests/ci/upload-coverage-report.sh:1]: No functional change was made. .travis.yml has all the stuff

File: tests/phpunit/MessageReporterAwareTest.php (hunk 3 above)
Comment (mwjames) [reply tests/phpunit/MessageReporterAwareTest.php]: > Because the test is not testing any production code logic and violates PSR-1 by having two classes in a file

Well, then is needs to be an integration test. The purpose of this exercise is that `MessageReporterAwareTrait` is tested to be working as expected (it should but tests provide clarity on the intent and behaviour).

File: tests/ci/upload-coverage-report.sh (hunk 4 above)
Comment (mwjames) [reply tests/ci/upload-coverage-report.sh:1]: +1

File: tests/phpunit/MessageReporterAwareTest.php (hunk 3 above)
Comment (JeroenDeDauw) [reply tests/phpunit/MessageReporterAwareTest.php]: Oh... I had not seen the trait. Probably due to a combination of me not being used to traits and always placing them on the first line in the class, not the second one.

**Comment by JeroenDeDauw** [comment]
Issues should be fixed now. I have use for the new CallbackMessageReporter, so hope there can be a release soon.

**Comment by mwjames** [comment]
> Issues should be fixed now. I have use for the new CallbackMessageReporter,
> so hope there can be a release soon.

Then I say, merge it and make a 1.4.0 release (we keep the .0 for a
symmetric aesthetic reason with other packages).

Cheers

**Comment by JeroenDeDauw** [comment]
Release made
